### PR TITLE
PartDesign: allow custom threading

### LINKED ISSUE #6533: PartDesign: Expose custom pitch and diameter of threaded hole
### Is there an existing request for this?

- [X] I have searched the existing requests

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?f=19&t=66805&p=577475#p577475

### Subproject(s) affected?

PartDesign

### Idea description

Starting from release 0.20, with the hole tool it is possibile to model real threads.
It would be useful to completely customize the thread, exposing editable pitch and diameter of the thread in the hole menu.

### Anything else?

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by benj5378**
Exposes thread diameter and thread pitch to property editor, for users being able to customize threads to their desire.
image

### CODE REVIEW

**Comment by benj5378** [comment]
@acominelli @simonachmueller @rjordans @petergalgano @cbabos @Anthony-Gaudino @antekone @botikori

Maybe you learned otherwise in school, but pestering actually works. @botikori proofed that by beginning to message me on Discord as well.

Can you guys please test this new feature?

**Comment by benj5378** [comment]
Fixes #6533

**Comment by benj5378** [comment]
> the order of thread type enums cannot be changed or you will break savefile compatibility (you will have to code a fixer) [@alfrix on Discord]

File: src/Mod/PartDesign/App/FeatureHole.cpp [hunk 1]
```diff
@@ -1400,16 +1353,25 @@ void Hole::onChanged(const App::Property* prop)
...
-            if (type != "None") {
+            if (type != "Custom") {
                 findClosestDesignation();
             }
         }
 
-        if (type == "None") {
+        if (type == "Custom") {
+            ThreadPitch.setReadOnly(false);
+            ThreadDiameter.setReadOnly(false);
```
Comment (alfrix) [inline src/Mod/PartDesign/App/FeatureHole.cpp:1363]: readOnly attributes are set below, try to keep it in one place please

**Comment by alfrix** [comment]
another issue is that you need to check the distance between minor and major diameters, because there is a range 

basically check that `Diameter > ThreadDiameter - 2 * maxHeight + marginZ`

marginZ i defined as 0.001 in the code, making it too small will cause OCCT to lag a lot, concident will make it crash in my machine

maxHeight for the base profile would be 7/8 of the sharpV
```
H = (sqrt(3) / 2 * Pitch) // height of sharpV
maxHeight = 7 *  H / 8;
```

File: src/Mod/PartDesign/App/FeatureHole.cpp (hunk 1 above)
Comment (benj5378) [reply src/Mod/PartDesign/App/FeatureHole.cpp:1363]: This was on purpose. The alternative was adding `setReadOnly(true)` in all the `else if` clauses

**Comment by benj5378** [comment]
> another issue is that you need to check the distance between minor and major diameters, because there is a range

I deliberately chose not do that, as it will cause UX issues:

Imagine 14 mm thread diameter, then hole diameter can be 14-12mm...

But you wan't to change diameter, so now you need to change thread diameter first. Then it won't be able to go lower than 12 mm. The hole diameter then needs to be changed again... going back and forth.

If the thread diameter could go lower than the 12 mm, the hole diameter could just follow along. But imagine someone putting the wrong thread diameter by mistake, and the their hole diameter is disrupted...

Not having the check just means, that it renders a bit weird. If you have any suggestions on how the check should be enforced, please let me know.

image

File: src/Mod/PartDesign/App/FeatureHole.cpp (hunk 1 above)
Comment (alfrix) [reply src/Mod/PartDesign/App/FeatureHole.cpp:1363]: move it below `bool isCustom = type == "Custom";`
as
```
ThreadPitch.setReadOnly(!(isCustom && isThreaded));
ThreadDiameter.setReadOnly(!(isCustom && isThreaded));
```

**Comment by alfrix** [comment]
it's ok then, as long as it doesn't cause a crash

do you want help with the TaskPanel changes?

**Comment by hyarion** [comment]
I'm getting an uneasy feeling of looking at this, from a maintainability perspective.

Adding support for this basically locks us into this specific implementation and migrating from it will be really difficult.
Too much customization exposed to the user makes it really difficult to change anything later on (with backwards compatibility), so great care needs to be taken when designing these power features.

I would prefer if custom threads are first implemented through UI.

**Comment by benj5378** [comment]
I'm not sure how setting some properties customly impacts maintainability? @hyarion

**Comment by hyarion** [comment]
> I'm not sure how setting some properties customly impacts maintainability? @hyarion

The problem isn't the properties, but to support the full range of possibilities and edge cases enabled by those properties.

**Comment by petergalgano** [comment]
Thanks for working on this. 

I would note that fully custom threading is one matter, but there are a number of ISO threading options that are not available to the user. I, for instance was unable to model M30x2.5, which do not correspond with "fine" (2mm) or "coarse" (3.5mm). I'm uncertain why, but for some reason, two shafts on one of our machines are threaded as such, and we were hoping to make models to remake the brass capture nuts that wear over time.

If you review lists of extended threading options for metric from engineering edge, you can see there are many many options that are available within the standard: https://www.engineersedge.com/hardware/metric-external-thread-sizes1.htm. The link is for external threads, includes options for different fits, Internal charts are also available, in sizing up to a remarkable M1060

https://burovik.biz/Metricthread.pdf Here is a extended subset of the standard...
https://www.bossard.com/-/media/bossard-group/website/documents/technical-resources/en/f-079-en.pdf lists the recommendations for selecting fine or coarse threading...

I presume that the modeling of the standard pitches is properly implemented, my hope is that I can gain the ability to model the less common ones, add the relevant parameters to the data.

Thanks again for those who are working on this.

**Comment by hyarion** [comment]
@maxwxyz [my comment here](https://github.com/FreeCAD/FreeCAD/pull/19801#issuecomment-2680982530) is still my standpoint
> > I'm not sure how setting some properties customly impacts maintainability? @hyarion
> 
> The problem isn't the properties, but to support the full range of possibilities and edge cases enabled by those properties.

Since FreeCAD holds backwards compatible very high we need to be really careful when increasing the scope of what we need to support so we don't add unnecessary maintenance cost when we want to modify the implementation later.

We are currrently planning to reimplement the treading support both in this feature and add a new threading tool. It would make more sense to me to add customization after those features so doing that work won't be affected by possible edge cases that we enable here.

**Comment by chennes** [comment]
@benj5378 is going to write some tests.

**Comment by jbaehr** [comment]
Changing the type "None" to "Custom" looks also strange from a user perspective. I mean, it's a "hole" tool, not a "thread" tool. So "None" for non-threaded holes sounds more natural to me.

Then the current implementation implies that "Custom" is an ISO Metric "M" / UTS profile (60°, ...). Yet we also have Whitworth threads (55°, ...). How to customise those?  
And Tomorrow we may also want to support some trapezoidal threads (e.g. for leadscrews) such as 30° ISO Metric "Tr" or 29° Acme. Here, too, soner or later someone will ask for customisation...

So I think leaving "None" for "no thread" and introducing additionally "Custom ISO M / UTS" and "Custom Whitworth", is the safer choice.

**Comment by alfrix** [comment]
Adding a combo for profile selection wouldn't be that hard, and I wouldn't consider it essential for an initial implementation.
Due to the way freecad saves this enums, I think the change from none to custom is an acceptable compromise

**Comment by jbaehr** [comment]
> Adding a combo for profile selection wouldn't be that hard

Do you mean an additional enum property for choosing between "ISO M / UTS" and "Whitworth"? This would again require quite some corner-case handling... Without additional checking/locking I could then choose e.g. an "M8x0.75" thread and a "Whitworth" profile. I fear that this will make the code yet again more complex -- a complexity that is not needed when making "Custom ISO M / UTS" and "Custom Whitworth" part of the Thread type enum instead of repurposing "None" to "Custom" (meaning "Custom ISO M / UTS" for now) and later introduce another property further refining "Custom".

**Comment by alfrix** [comment]
> > Adding a combo for profile selection wouldn't be that hard
> 
> Do you mean an additional enum property for choosing between "ISO M / UTS" and "Whitworth"? This would again require quite some corner-case handling... Without additional checking/locking I could then choose e.g. an "M8x0.75" thread and a "Whitworth" profile. I fear that this will make the code yet again more complex -- a complexity that is not needed when making "Custom ISO M / UTS" and "Custom Whitworth" part of the Thread type enum instead of repurposing "None" to "Custom" (meaning "Custom ISO M / UTS" for now) and later introduce another property further refining "Custom".

No, Custom as implemented here wouldn't have a size selector, you need to input your size manually, take a look a my UI Proposal for this: https://github.com/FreeCAD/FreeCAD/pull/19839

In the case of a profile selector it would look like this (only the first 2 are implemented currently):
image
and that selector would only available on Custom

I want to address something:
**If what is preferred is to just customize a standard, then that would be different than what this PR is offering.**

In that case we should not modify the None profile, and add a "Custom" checkbox that hides the Size combo and enables the modifications

And that is the fundamental question:
Should the None Profile be allowed to be threaded (as in this PR), or should a custom checkbox be added to modify an existing profile?

**Comment by petergalgano** [comment]
I looked at your UI proposals, they look great, very complete. I'm sure the results will be very helpful.

My personal interest is generally in being able to add sizes/pitches from the metric standards that are missing in the current selections whether that were in a separate profile or not.. I'm presuming that there is a database of the parameters for the selections that could be edited. I did look for such a thing in the files of the install, and didn't find something that I could hack.

There are so many parameters, including fits and tolerances, I'm sure it takes a fair amount to implement this fully, but an interim step of being able to add individual threads that come up would be appreciated by me.

**Comment by jbaehr** [comment]
@alfrix thanks for your explanations and examples. Now it makes much more sense for me. 👍

**Comment by luzpaz** [comment]
Any chance this can get resurrected for 1.2 ?

Also related: https://github.com/FreeCAD/FreeCAD/pull/22573
